### PR TITLE
[DS][31/n] Create static constructor for eager condition

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

This creates a static constructor for an analog to the current-day eager condition. It is not identical, as it does not consider historical time partitions. It also manages state differently.

In the current world, the "parent_newer" analog actually encodes a lot of different information. From its perspective, an asset partition is True if:

A parent has been materialized more recently than the asset, and the run that materialized it does not target this asset,
AND this asset hasn't been requested for materialization by AMP since that run came in

This is a pretty confusing thing to wrap your head around, but it does have two good qualities:

it avoids repeatedly requesting the downstream asset for materialization while the asset computes (techincally, the parent will remain newer than the child until the child successfully materializes, which may in practice span a large number of scheduling evaluations)
it avoids repeatedly materializing the asset in the case that the requested materialization failed (similar to the above issue)
This policy handles issue (1) by explicitly avoiding materializing this asset if a run is currently in progress, but does not handle case (2).

This will need to be handled with a recently_requested condition, which can keep track of the asset partitions which have been requested within some timedelta.

### CODE REVIEW

**Comment by OwenKephart** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21737?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#21712** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21712?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21705** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21705?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21704** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21704?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21703** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21703?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21677** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21677?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21737** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21737?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#21671** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21671?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21641** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21641?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#21670](https://github.com/dagster-io/dagster/pull/21670) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21670?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#21640** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21640?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21648** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21648?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21615** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21615?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21613** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21613?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21612** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21612?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21573** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21573?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21546** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21546?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21545** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21545?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21541** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21541?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21540** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21540?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21539** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21538** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21538?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21537** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21537?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21536** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21536?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21535** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21535?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21521** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21521?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21520** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21520?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21511** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21511?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#21512](https://github.com/dagster-io/dagster/pull/21512) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21512?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#21510** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21510?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21508** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21508?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21507** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21507?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21505** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21505?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21504** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21504?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21503** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21503?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21502** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21502?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21501** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21501?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21500** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21500?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21499** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21499?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/declarative_scheduling/scheduling_condition.py [hunk 1]
```diff
@@ -147,6 +147,40 @@ def parent_newer() -> "ParentNewerCondition":
...
+    @staticmethod
+    def eager() -> "SchedulingCondition":
+        """Returns a condition which will "eagerly" fill in missing partitions as they are created,
+        and ensures unpartitioned assets always contain the latest available data.
+
+        Specifically, this is a composite SchedulingCondition which is true for an asset partition
+        if all of the following are true:
+
+        - The asset partition is within the latest time window
+        - At least one of its parents has been materialized more recently than the asset partition,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/declarative_scheduling/scheduling_condition.py:159]: Should we say updated to be inclusive of observable stuff?

File: python_modules/dagster/dagster/_core/definitions/declarative_scheduling/scheduling_condition.py [hunk 2]
```diff
@@ -147,6 +147,40 @@ def parent_newer() -> "ParentNewerCondition":
 
         return ParentNewerCondition()
 
+    @staticmethod
+    def eager() -> "SchedulingCondition":
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/declarative_scheduling/scheduling_condition.py:151]: I believe it does, but can we explicitly state that an `eager` scheduling condition will kick off if one of its parents is the target of an _ad hoc_ update?

**Review by schrockn — CHANGES_REQUESTED** [review]
Super exciting to see this. Just want to make sure I understand 100% our default policy here and have it in the docblock

**Review by schrockn — APPROVED** [review]
Please add test that ensures this actually constructs. 

All I am looking for is

`assert isinstance(SchedulingCondition.eager(), SchedulingCondition)`

**Comment by OwenKephart** [comment]
a test gets added upstack, which also tests the behavior of the composite condition

**Comment by OwenKephart** [comment]
### Merge activity

* **May 14, 10:38 AM EDT**: @OwenKephart started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/21737).

**Pull request merged**